GenericList/string_list.cpp

```diff
@@ -4,47 +4,96 @@
 
-#include "string_list.h"
+#include "string_list.h"'
+
+std::allocator<std::string> List_String::dataAllocator;
 
 List_String::List_String() : data(nullptr), capacity(0), count(0) {
 
 }
 
+
+void swap(List_String& first, List_String& second) {
+	//explanation: calling swap makes it unqualified, meaning if swap(T,T) is specially defined, it is called, otherwise, call std::swap(T,T)
+	swap(first.capacity, second.capacity);
+	swap(first.count, second.count);
+	swap(first.data, second.data);
+}
+//Rule of 3
+
+//Destructor
 List_String::~List_String() {
-	delete[] data;
+	for (int i = 0; i < count; i++) {
+		data[i].~string();
+	}
+	dataAllocator.deallocate(data, capacity);
 }
 
-int List_String::Capacity() { return capacity; }
+//Copy constructor
+List_String::List_String(const List_String& other) : capacity(other.capacity), count(other.count) {
+	data = CreateDeepCopy(other.data, other.capacity, other.capacity);
+}
 
-bool List_String::Capacity(int new_capacity) {
-	if (new_capacity < 0) return false; //allowed values must be >= 0
-	if (new_capacity < count) return false; //can't have shorter capacity than element count
-	if (new_capacity == capacity) return true; //no change
-	if (new_capacity == 0) { //empty non-empty array
-		delete data;
-		return true;
-	};
+//Copy assignement
+List_String& List_String::operator=(const List_String& other) {
+	//Note: other is not a const reference but a copy of the value, allowing for use of swap logic (since "other" will be destroyed after
+	//Long definition:
+		//List_String tmp(other);
+		//swap(*this, tmp);
+		//return *this;
+	//clearer definition:
+	return *this = List_String(other); //calls operator=(&& List_String(&)))
 	
-	// general case: dynamic resizing
-	std::string* new_data = new std::string[new_capacity]; 
-	for (int i = 0; i < capacity; i++) {
-		new_data[i] = data[i];
+}
+
+
+//Rule of 5
+//R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+//basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+List_String::List_String(List_String&& other) : data(nullptr), capacity(0), count(0) {
+	swap(*this, other);
+}
+
+List_String& List_String::operator=(List_String&& other) {
+	//warning for future use: might need to do additional cleaning here if currently controlling external objects (since control will be passed over to other and deleted)
+	swap(*this, other);
+	return *this;
+}
+
+
+
+int List_String::Capacity() const { return capacity; }
+
+void List_String::Capacity(int new_capacity) {
+	if (new_capacity < 0) return; //allowed values must be >= 0
+	if (new_capacity < count) return; //can't have shorter capacity than element count
+	if (new_capacity == capacity) return; //no change
+	if (new_capacity == 0) { //empty non-empty array with no data
+		dataAllocator.deallocate(data, capacity); //no need to destruct anything since no constructed data exists
+		data = nullptr;
```

Comment:
You _do_ need to destroy all the elements, constructed data can exist
```suggestion
		Clear();
		dataAllocator.deallocate(data, capacity);
		data = nullptr;
```